### ISSUE TITLE
Hang in firehose section

### ISSUE THREAD

**Issue by z3ntu** [bkerler/edl#59]
Hi, I'm trying your tool on a msm8974 device where I should have a working loader for
Unfortunately it hangs somewhere in the firehose section during a usb read, here's the log:
```
main - Device detected :)
main - Mode detected: sahara
Device is in EDL mode .. continuing.
sahara - 
------------------------
HWID:              0x007bc0e100000000 (MSM_ID:0x007bc0e1,OEM_ID:0x0000,MODEL_ID:0x0000)
CPU detected:      "MSM8974ABv3"
PK_HASH:           0xcc3153a80293939b90d02d3bf8b23e0292e452fef662c74998421adad42a380f
Serial:            0x108211e7

sahara - Uploading loader /tmp/MPRG8974.mbn ...
sahara - Loader uploaded.
Successfully uploaded programmer :)
firehose
firehose - [LIB]: Nop failed.
firehose - No supported functions detected, configuring qc generic commands
firehose - 
firehose_client - Target detected: MSM8974ABv3
firehose_client - Based on the chipset, we assume eMMC as default memory type...
firehose - TargetName=
firehose - MemoryName=eMMC
firehose - Version=


^CTraceback (most recent call last):
  File "/tmp/edl/./edl.py", line 360, in <module>
    base.run()
  File "/tmp/edl/./edl.py", line 338, in run
    fh = firehose_client(args, self.cdc, self.sahara, self.__logger.level, print)
  File "/tmp/edl/Library/firehose_client.py", line 89, in __init__
    if self.firehose.configure(0):
  File "/tmp/edl/Library/firehose.py", line 891, in configure
    rsp = self.cmd_read_buffer(0, 1, 1, False)
  File "/tmp/edl/Library/firehose.py", line 674, in cmd_read_buffer
    tmp = self.cdc.read(self.cfg.MaxPayloadSizeToTargetInBytes)
  File "/tmp/edl/Library/usblib.py", line 331, in read
    tmp = self.device.read(self.EP_IN, length, timeout)
  File "/usr/lib/python3.9/site-packages/usb/core.py", line 1019, in read
    ret = fn(
  File "/usr/lib/python3.9/site-packages/usb/backend/libusb1.py", line 846, in bulk_read
    return self.__read(self.lib.libusb_bulk_transfer,
  File "/usr/lib/python3.9/site-packages/usb/backend/libusb1.py", line 946, in __read
    retval = fn(dev_handle.handle,
KeyboardInterrupt
```
I've Ctrl-C'ed the program after waiting for multiple minutes. I've tried with multiple functions like `printgpt`, `getstorageinfo` or `qfil`, all show the same behavior.

**Comment by kercre123** [comment]
Have you tried other tools to verify this loader works? Maybe try different USB ports, USB hubs, and cables.

**Comment by z3ntu** [comment]
I haven't had any luck with QFIL (the .exe you can find on the internet - that failed with `Download Fail:FireHose Fail:FHLoader Fail:The system cannot find the file specified` where I'm not sure what it's looking for)

https://github.com/openpst/sahara This seem to be able to upload the loader fine but that tool can't do firehose (or the qfil dance).

I'm also not particularly sure how or if the `prog_emmc_firehose_8974.mbn` file is being used by this tool; QFIL (windows .exe) seems to only use this and not the `MPRG8974.mbn` or `sahara.xml`

**Comment by z3ntu** [comment]
Actually I've just managed to flash the device with QFIL 2.0.0.0 (it was just 2.0.3.5 that didn't work with this error), so the loader should definitely be fine.

**Comment by bkerler** [comment]
@z3ntu Luca, can you retry with --debugmode and send me the log.txt ? Thx

**Comment by z3ntu** [comment]
Hi, here's the log: [edl_py_log.txt](https://github.com/bkerler/edl/files/6031580/edl_py_log.txt)
I've removed most of the loader payload as I'm not sure I should share this publicly.

**Comment by bkerler** [comment]
I think that you're using the wrong loader. The loader doesn't respond to firehose commands according to the log, so it's very likely you are using the streaming mode instead with that loader. That might be why only old QFIL support this mode.
As the device is unfused, you might search and use other loaders which might just work fine for your cpu.

**Comment by z3ntu** [comment]
Do you know what the Windows QFIL application uses? As far as I understand it, Sahara is a protocol with which the first loader (the MPRG8974.mbn file) is uploaded and validated; this is expected to provide a firehose protocol interface then for this program. Is qfil another protocol that's normally started from firehose?

**Comment by bkerler** [comment]
sahara is the first stage on hardware level. Depending on the loader you send, it's either dload streaming or firehose. The newer devices all have firehose loaders. The streaming format is based on hdlc codes, whereas the firehose is based on xml. I've added basic streaming support but only for nand devices. MPRG is normally used for download streaming. You should normally use ENPRG for sahara instead for streaming emergency, or firehose_emmc / firehose_ufs for firehose.

**Comment by z3ntu** [comment]
In the QFIL Windows application I selected "Flat build" and added the rawprogram0.xml and patch0.xml files and `prog_emmc_firehose_8974.mbn` as programmer programmer. Looks similar to https://repairmymobile.in/wp-content/uploads/2020/02/flat-1.png . I guess there is some firehose stuff in there though given the programmer filename?
Anyways if you think this is not a bug in edl and you don't want to support this feel free to close the issue, I don't have an immediate use case for this anymore (even though it would be nice to flash this with open source software) :D

**Comment by bkerler** [comment]
What device model / vendor is this ? I will have a look if I have any msm8974 device floating around

**Comment by bkerler** [comment]
I've just tried with an oneplus X (msm8974AB) and it's working fine. So I think the problem really might be the firehose programmer. Can you upload that programmer ?

**Comment by tzkrn** [comment]
Hello, @bkerler.
I have a similar issue with nexus 6 (shamu). The programmer is supposed to be correct and the sahara protocol looks like it's working. However, it fails in a similar way as @z3ntu described (the debug log is pretty similar with the loop in 'Timed out'.
Example:

  ```
$ ./edl.py reset
  Qualcomm Sahara / Firehose Client V3.2 (c) B.Kerler 2018-2021.
  main - Trying with no loader given ...
  main - Waiting for the device
  main - Device detected :)
  main - Mode detected: sahara
  Device is in EDL mode .. continuing.
  sahara - 
  ------------------------
  HWID:              0x009010e102280000 (MSM_ID:0x009010e1,OEM_ID:0x0228,MODEL_ID:0x0000)
  Unknown CPU, please send log as issue to https://github.com/bkerler/edl
  PK_HASH:           0xf541e6985d979cdd829ff504494b6ebfafc9c88a1329d36c1948bf9aa22902b4
  Serial:            0x00dba6e4
  
  sahara - Detected loader: Loaders/009010e102280000_f541e6985d979cdd_FHPRG.bin
  sahara - Uploading loader Loaders/009010e102280000_f541e6985d979cdd_FHPRG.bin ...
  sahara - Loader uploaded.
  Successfully uploaded programmer :)
  firehose
  firehose - [LIB]: Nop failed.
  firehose - No supported functions detected, configuring qc generic commands
  firehose - 
  firehose_client - No --memory option set, we assume "eMMC" as default ..., if it fails, try using "--memory" with "UFS","NAND" or "spinor" instead !
  firehose - TargetName=
  firehose - MemoryName=eMMC
  firehose - Version=

... (hang) ...
```
----
By the way, there is something I don't fully understand and might be the reason why this is not working. In [1] this device is listed along with its programmer, but it's described as non-Firehose. What does it mean? 

Thank you for your time.


Nexus 6 (shamu) - APQ8084 - programmer.mbn (non-Firehose)
9b3184613d694ea24d3beeba6944fdb64196fea7056c833d38d2ef683fd96e9b  programmer.mbn
[1] https://alephsecurity.com/2018/01/22/qualcomm-edl-1/

**Comment by bkerler** [comment]
Yes , that's right. It's a streaming download mode, which so far I only partly added support for. Hopefully I can add support for it soon.

**Comment by tzkrn** [comment]
Alright I understand. Thank you, @bkerler !

**Comment by bkerler** [comment]
Regarding the original issue, please try commit ff372c3

**Comment by gitquelm** [comment]
Hey All,

This is an old issue, but I'm hoping someone is still watching. I'm having a similar issue where EDL hangs at:

C:\edl>python edl getstorageinfo --portname \\.\COM3
Qualcomm Sahara / Firehose Client V3.61 (c) B.Kerler 2018-2023.
main
main - [LIB]: ←[33mPlease first install libusb_win32 driver from Zadig←[0m
main - Trying with no loader given ...
main - Waiting for the device
main - Device detected :)
DeviceClass - Warning !
main - Mode detected: sahara

I'm trying to get a stock Samsung S21 (G991U) out of EDL mode. I'm running Windows. I get Windows to detect the 9008 driver just fine. I can also use Zadig to change the driver to libusb. That all works fine too. But I can't get the phone to respond to commands or reset with EDL. Tried all the combinations of vol+-/power. I have a black screen and it won't do anything else but sit in 9008 mode. The phone froze out of nowhere tonight while I was using it. It's been fine the whole time I've had it. Like I mentioned, it's completely stock. Never any custom firmware. After being frozen for a few minutes the screen went black and here we are. I'd really like to get some data from this phone before attempting to wipe or flash it. Any help would be appreciated!

Note that if I leave the com port driver active it hangs. If I change to libusb, EDL appears to complete, but nothing happens to the phone. I've tried getstorageinfo and reset commands. Do I need a proper loader for this phone?

C:\edl>python edl getstorageinfo
Qualcomm Sahara / Firehose Client V3.61 (c) B.Kerler 2018-2023.
main
main - [LIB]: ←[33mPlease first install libusb_win32 driver from Zadig←[0m
main - Trying with no loader given ...
main - Waiting for the device
main - Device detected :)

C:\edl>
